dbms/src/Storages/KVStore/Utils/AsyncTasks.h

```diff
@@ -14,110 +14,358 @@
+            // it must have been registered into `tasks`.
+            // So the only case that an access for a non-existing task is that the task is already cancelled asyncly.
+            return CancelHandle::genAlreadyCanceled();
+        }
+        return it->second.cancel;
+    }
+
+    // Only unregister, no clean.
+    // Use `asyncCancelTask` if there is something to clean.
+    bool leakingDiscardTask(Key k)
     {
         std::scoped_lock l(mtx);
-        auto it = futures.find(k);
-        if (it != futures.end())
+        auto it = tasks.find(k);
+        if (it != tasks.end())
         {
-            futures.erase(it);
-            start_time.erase(k);
+            tasks.erase(it);
             return true;
         }
         return false;
     }
 
-    bool addTask(Key k, Func f)
+    // Safety: Throws if
+    // 1. The task not exist and `throw_if_noexist`.
+    // 2. Throw in `result_dropper`.
+    /// Usage:
+    /// 1. If the task is in `Finished` state
+    ///     It's result will be cleaned with `result_dropper`.
+    /// 2. If the task is in `Running` state
+    ///     Make sure the executor will do the clean.
+    /// 3. If the tasks is in `InQueue` state
+    ///     The task will directly return when it's eventually run by a thread.
+    /// 4. If the tasks is in `NotScheduled` state
+    ///     `throw_if_noexist` controls whether to throw.
+    /// NOTE: The task element will be removed after calling this function.
+    template <typename ResultDropper>
+    TaskState asyncCancelTask(Key k, ResultDropper result_dropper, bool throw_if_noexist)
+    {
+        auto cancel_handle = getCancelHandleFromCaller(k, throw_if_noexist);
+        if (cancel_handle)
+        {
+            cancel_handle->doCancel();
+            // Cancel logic should do clean itself
+        }
+
+        auto state = queryState(k);
+        if (!throw_if_noexist && state == TaskState::NotScheduled)
+            return state;
+        if (state == TaskState::Finished)
+        {
+            result_dropper();
+        }
+
+        // `result_dropper` may remove the task by `fetchResult`.
+        leakingDiscardTask(k);
+
+        return state;
+    }
+
+    TaskState asyncCancelTask(Key k)
     {
+        return asyncCancelTask(
+            k,
+            []() {},
+            true);
+    }
+
+    // Safety: Throws if
+    // 1. The task is not found, and throw_on_no_exist.
+    // 2. The cancel_handle is already set.
+    // 3. Throw in `result_dropper`.
+    /// Usage:
+    /// 1. If the task is in `Finished`/`Running` state
+    ///     It's result is returned. The Caller may do the rest cleaning.
+    /// 3. If the tasks is in `InQueue` state
+    ///     The task will directly return when it's eventually run by a thread.
+    /// 4. If the tasks is in `NotScheduled` state
+    ///     It will throw.
+    /// Returns:
+    /// 1. `NotScheduled` or `InQueue`
+    /// 2. `R`
```

Comment:
Is it `Running` or `Finished`?

Reply:
No. `blockedCancelRunningTask` means block cancel